modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexParallelizationHelper.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.reindex;
+
+import org.elasticsearch.action.Action;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.util.concurrent.AtomicArray;
+import org.elasticsearch.index.mapper.UidFieldMapper;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.tasks.TaskManager;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ReindexParallelizationHelper {
+    public static <
+                Request extends AbstractBulkByScrollRequest<Request>
+            > void startSlices(NodeClient client, TaskManager taskManager, Action<Request, BulkIndexByScrollResponse, ?> action,
+                    String localNodeId, ParentBulkByScrollTask task, Request request, ActionListener<BulkIndexByScrollResponse> listener) {
+        /* Explicitly tell the task manager that we're running child tasks on the local node so it will cancel them when the parent is
+         * cancelled. */
+        taskManager.registerChildTask(task, localNodeId);
```

Comment:
Yes, it is required but the placement of it looks really weird. We typically register child tasks just before dispatching a request to another node. It basically lets the taskmanager know where to search for children. In this case all requests are executed locally, therefore it's possible to register it only once.